config/server.py

```diff
@@ -0,0 +1,196 @@
+import uuid
+import argparse
+from urllib.parse import urlparse
+from functools import wraps
+
+import connexion
+import flask
+from google.oauth2 import id_token
+from google.auth.transport import requests
+from peewee import CharField, Model, SqliteDatabase
+
+
+def get_route_model(db):
+    class Route(Model):
+        owner = CharField()
+        name = CharField()
+        destination = CharField()
+        token = CharField()
+
+        def get_json(self):
+            return {
+                "owner": self.owner,
+                "destination": self.destination,
+                "token": self.token,
+                "name": self.name
+            }
+
+        class Meta:
+            database = db
+
+    return Route
+
+
+class InvalidCredentials(Exception):
+    pass
+
+
+class NotAuthorised(Exception):
+    pass
+
+
+class InvalidRouteID(Exception):
+    pass
+
+
+class InvalidURL(Exception):
+    pass
+
+google_oauth_clientID = "859663336690-q39h2o7j9o2d2vdeq1hm1815uqjfj5c9.self.apps.googleusercontent.com"
+
+class Auth:
+    def get_user(self):
+        token = flask.request.headers.get("Google-Auth-Token")
+
+        if token is None:
+            raise InvalidCredentials()
+
+        try:
+            token_info = id_token.verify_oauth2_token(token, requests.Request(), google_oauth_clientID)
+        except ValueError as ve:
+            raise InvalidCredentials() from ve
+
+        if token_info["hd"] != "sanger.ac.uk":
+            raise InvalidCredentials()
+
+        if token_info['iss'] not in ['accounts.google.com', 'https://accounts.google.com']:
+            raise InvalidCredentials()
+
+        return token_info["email"]
+
+class Server:
+    def _auth_request(func):
+        @wraps(func)
+        def new_func(self, *args, **kw_args):
+            current_user = self.auth.get_user()  # includes auth
+
+            self._log_function_call(func.__name__, current_user, args)
+
+            return func(self, *args, **kw_args)
+
+        return new_func
+
+    def _log_function_call(self, name, user, parameters):
+        # TODO
+        pass
+
+    def _token2route(self, token: str) -> get_route_model(None):
```

Comment:
`get_route_model` does return a type, so I can kind of see where you're going with this. I'm not sure if this -- effectively [dependent typing](https://en.wikipedia.org/wiki/Dependent_type) -- is allowed/supported in the grand scheme of Python type hinting. A better approach might be for your `get_route_model` to subclass an abstract type, then use that as the return type; especially since you're passing `None` to your factory.

Reply:
This is valid python type - the type checker just works out the type of `get_route_model(None)` (which is `Route`) and makes it the return type of this function. I've tried subclassing another class which defines the properties, but this made internal peewee code error, so I've extracted the Route type as follows:
```python
def _helper() -> get_route_model:
    return None

Route = _helper()
```

Reply:
Is "peewee" some kind of ORM? Can you not do something like:

```python
from abc import ABCMeta, abstractmethod

class AbstractBaseRoute(Model, metaclass=ABCMeta):
    @abstractmethod
    def get_json(self):
        """ Docstring for method """

def get_route_model(db):
    class Route(AbstractBaseRoute):
        owner = CharField()
        name = CharField()
        destination = CharField()
        token = CharField()

        def get_json(self):
            return {
                "owner": self.owner,
                "destination": self.destination,
                "token": self.token,
                "name": self.name
            }

        class Meta:
            database = db

    return Route
```

...then you can use `AbstractBaseRoute` as the return type annotation, wherever you need it. Dynamically creating types, solely for the type checker's benefit, at runtime seems like too high a hoop to jump through. Type annotations should aid a human's understanding as much as an automated tool's with the least amount of surprise. While your method currently works, it does so at the expense of "obviousness".

Reply:
As far as I can determine, there is no provision for dependent types in PEP 484. However, it is syntactically correct as per PEP 3107 and therefore it is possible that an analyser _could_ interpret the return as a dependent type. The analyser I'm using, for example, does not acknowledge it though.

The `BaseRoute` models Chris suggests is the best solution here. Everything minus the `Meta` subclass can be pulled up into that.